### PR TITLE
capture http call related errors

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
Report is now able to show/distinct http calls with error (e.g. wrong url protocol).
TestStep refactoring.

### CODE REVIEW

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestStep.java [hunk 1]
```diff
@@ -45,7 +45,7 @@
     public static <E> TestStep<E> create(E context,
                                          TokenizedMessage inProgressMessage,
                                          Supplier<TokenizedMessage> completionMessageSupplier,
-                                         Runnable action) {
+                                         Supplier action) {
```
Comment (tsiq-karold) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/TestStep.java]: Presumably this supplier always supplies an R?  If so, couldn't we parameterise TestStep on it?  Not really fond of using generic classes with no generic params (not least because IntelliJ moans at you for it 😄 )

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java [hunk 2]
```diff
@@ -29,34 +31,35 @@
     public static <C, V> void shouldStep(C context, V value, StepReportOptions stepReportOptions, TokenizedMessage valueDescription, ValueMatcher valueMatcher) {
         executeStep(context, value, valueDescription, valueMatcher, false,
                 tokenizedMessage(action("expecting")),
-                () -> actual(value).should(valueMatcher), stepReportOptions);
+                runnable(() -> actual(value).should(valueMatcher)), stepReportOptions);
```
Comment (tsiq-karold) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java]: Looks like all invocations of `executeStep` take what is actually a `Runnable`.  Therefore, I think we should either have `executeStep` take a `Runnable` and convert to a `Supplier` (rather than doing the conversion in each `executeStep` invocation site) or we create an overload of `TestStep.create` which takes a `Runnable` and converts it.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java [hunk 3]
```diff
@@ -69,4 +72,11 @@
 
         step.execute(stepReportOptions);
     }
+
+    private static Supplier runnable(Runnable r) {
```
Comment (tsiq-karold) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java]: This could return `Supplier<Void>`

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/stacktrace/StackTraceUtils.java [hunk 4]
```diff
@@ -113,11 +113,9 @@ private static String filterStackTrace(Throwable t, Predicate<String> filter) {
         String stackTrace = renderStackTrace(t);
         List<String> lines = Arrays.asList(stackTrace.split("\n"));
 
-        return lines.get(0) + "\n" +
-                lines.subList(1, lines.size())
-                        .stream()
-                        .filter(filter)
-                        .collect(Collectors.joining("\n"));
+        return lines.stream()
```
Comment (tsiq-karold) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/stacktrace/StackTraceUtils.java]: Why this change?  I assumed we had the lines.get(0) because we didn't want to apply filtration on the message part of the stack trace and now that's no longer the case.  It is of course possible my assumption was incorrect...

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java (hunk 2 above)
Comment (tsiq-karold) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java]: Having looked through the rest of the review, I'd go for the `TestStep.create` overload as there are other classes that wrap a `Runnable` into a `Supplier` in exactly the same way.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestStep.java (hunk 1 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/TestStep.java]: I thought to add it, the problem is the actual type we only now at the `step.execute()` time and not the create time.

```
ThisIsTheType id = http.post(...) { 
    return id
}
```

so we can define it as a generic, but it won't do any good. Of course I may be missing some generic magic knowledge.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java (hunk 2 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/ValueMatcherExpectationSteps.java]: Makes sense. Thanks.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/stacktrace/StackTraceUtils.java (hunk 4 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/stacktrace/StackTraceUtils.java]: For one liners it was doing extra '\n'. Then I looked at what filters are being passed. We always check now for the presence of `at` part, which never happens on first line.
This filter function is private and we control use cases. That was justification.

Comment (tsiq-karold) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/stacktrace/StackTraceUtils.java]: Makes sense
